src/Controller/Component/AuthComponent.php

```diff
@@ -880,7 +880,7 @@ public function constructAuthenticate()
      *
      * @param \Cake\Auth\Storage\StorageInterface|null $storage Sets provided
      *   object as storage or if null returns configured storage object.
-     * @return \Cake\Auth\Storage\StorageInterface|\Cake\Core\InstanceConfigTrait|null
```

Comment:
Removing `\Cake\Core\InstanceConfigTrait` makes `phpstan` unhappy.
https://travis-ci.org/cakephp/cakephp/jobs/356494536#L732

Reply:
This is *definitely* an off-label use of traits, treating them like interfaces. PHP won't allow traits as return types, and I don't see why they should be allowed in a union type.

Reminds me of [this RFC](https://wiki.php.net/rfc/traits-with-interfaces), except without the interface.

I can add `@method` annotations to `StorageInterface` to make PHPStan happy.

Reply:
I am happy with whatever solution which makes both `phpstan` and `psalm` happy.

Mind you travis runs phpstan at level 2 but we are slowing working towards having it pass level 3. I hope the changes you make for `psalm` don't increases the errors reported for phpstan level 3. :slightly_smiling_face:

Reply:
> I hope the changes you make for psalm don't increases the errors reported for phpstan level 3

I can't imagine how they would.